### PR TITLE
Ports Nooses to bee so you can hang people

### PR DESCRIPTION
**Pull request by ro5490**
## About The Pull Request
Ported from Hippie - https://github.com/HippieStation/HippieStation/pulls?q=is%3Apr+noose

Adds nooses, not much else to say

## Why It's Good For The Game
Hang people for their crimes!

## Changelog
:cl:
add: Added Nooses
soundadd: added sound files for the noose
imageadd: added icons for the noose
fix: Take a cable coil, stand on a chair, build a noose, then drag whoever onto it to hang them
/:cl:

### CODE REVIEW

**Comment by mehmetipekpatates** [comment]
epic

File: code/game/objects/noose.dm [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/obj/item/stack/cable_coil/building_checks(datum/stack_recipe/R, multiplier)
+	if(R.title == "noose")
```
Comment (ike709) [inline code/game/objects/noose.dm]: Bad.

File: code/game/objects/noose.dm [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		if(!(locate(/obj/structure/chair) in usr.loc) && !(locate(/obj/structure/bed) in usr.loc) && !(locate(/obj/structure/table) in usr.loc) && !(locate(/obj/structure/toilet) in usr.loc))
```
Comment (ike709) [inline code/game/objects/noose.dm]: Bad.

**Comment by ike709** [comment]
Didn't have time to fully review.

I'm *sure* nothing bad will come of this.

**Comment by ro5490** [comment]
> Didn't have time to fully review.
> 
> I'm _sure_ nothing bad will come of this.

Not exactly sure whats bad about these, so if someone could explain I'll happily try and optimise/fix it

File: code/game/objects/noose.dm (hunk 1 above)
Comment (ro5490) [reply code/game/objects/noose.dm]: Not sure what exactly about this section of code is bad
`/obj/item/stack/proc/building_checks(datum/stack_recipe/R, multiplier)` is used in stack.dm already and this just specifies the cable_coil checks
and `if(R.title =="noose")` seems alright to me for defining the noose recipe

File: code/game/objects/noose.dm (hunk 2 above)
Comment (ro5490) [reply code/game/objects/noose.dm]: `		if(!(locate(/obj/structure/chair) || (/obj/structure/table) || (/obj/structure/toilet) in usr.loc))`
Better?

Comment (ro5490) [reply code/game/objects/noose.dm]: `		if(!(locate(/obj/structure/chair in usr.loc)))`
bit better?

Comment (ro5490) [reply code/game/objects/noose.dm]:
```
/obj/item/stack/cable_coil/building_checks(datum/stack_recipe/R, multiplier)
	if(R.title == "noose")
		if(!(locate(/obj/structure/chair in usr.loc)))
			to_chat(usr, "<span class='warning'>You have to be standing on top of a chair to make a noose!</span>")
			return FALSE
	return ..()
```

Comment (ro5490) [reply code/game/objects/noose.dm]:
```
/obj/item/stack/cable_coil/building_checks(datum/stack_recipe/R, multiplier)
	if(R.title == "noose")
		if(!(locate(/obj/structure/chair) in usr.loc))
			to_chat(usr, "<span class='warning'>You have to be standing on top of a chair to make a noose!</span>")
			return FALSE
	return ..()

```
bit better?

Comment (hotimep) [reply code/game/objects/noose.dm]: toliet noose

File: code/game/objects/noose.dm (hunk 1 above)
Comment (ike709) [reply code/game/objects/noose.dm]: String comparisons are slow. Check if `result_type` is the noose typepath instead.

File: code/game/objects/noose.dm [hunk 3]
```diff
@@ -0,0 +1,153 @@
+/obj/item/stack/cable_coil/building_checks(datum/stack_recipe/R, multiplier)
+	if(R.title == "noose")
+		if(!(locate(/obj/structure/chair) in usr.loc))
```
Comment (ike709) [inline code/game/objects/noose.dm]:
```suggestion
		if(!(locate(/obj/structure/chair) in get_turf(usr)))
```

File: code/game/objects/noose.dm [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	return ..()
+
+/obj/structure/chair/noose //It's a "chair".
+	name = "noose"
+	desc = "Well this just got a whole lot more morbid."
+	icon_state = "noose"
+	icon = 'icons/obj/objects.dmi'
+	layer = FLY_LAYER
+	flags_1 = NODECONSTRUCT_1
+	var/image/over
```
Comment (ike709) [inline code/game/objects/noose.dm]: Why is this not a `mutable_appearance`?

Comment (ro5490) [reply code/game/objects/noose.dm]: Which line? I'm still not too familiar with byondcode tbh

File: code/game/objects/noose.dm [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+										"<span class='suicide'>[buckled_mob]'s hands are desperately clutching the noose.</span>",\
+										"<span class='suicide'>[buckled_mob]'s limbs sway back and forth with diminishing strength.</span>")
+				if(buckled_mob.stat == DEAD)
+					flavor_text = list("<span class='suicide'>[buckled_mob]'s limbs lifelessly sway back and forth.</span>",\
+										"<span class='suicide'>[buckled_mob]'s eyes stare straight ahead.</span>")
+				buckled_mob.visible_message(pick(flavor_text))
+				playsound(buckled_mob.loc, 'sound/effects/noose_idle.ogg', 30, 1, -3)
+
+/mob/living/carbon/human
+	var/noosed = FALSE
```
Comment (ike709) [inline code/game/objects/noose.dm]: No. We don't need a `human` var to see if someone is being hanged. Just check if they're in the structure's `buckled_mobs` or not to keep track of this.

File: code/modules/surgery/bodyparts/head.dm [hunk 6]
```diff
@@ -141,6 +141,7 @@
 	else if(!animal_origin)
 		var/mob/living/carbon/human/H = C
 		var/datum/species/S = H.dna.species
+		H.checknoosedrop()
```
Comment (ike709) [inline code/modules/surgery/bodyparts/head.dm]: Why are you calling this in update_limb()?

File: code/game/objects/noose.dm (hunk 1 above)
Comment (ro5490) [reply code/game/objects/noose.dm]: `if(R.result_type(noose)) ?`
cant find many references to `result_type` in the code, only in stacks.dm

File: code/modules/power/cable.dm [hunk 7]
```diff
@@ -9,6 +9,11 @@ GLOBAL_LIST_INIT(cable_colors, list(
 	"red" = "#ff0000"
 	))
 
+
+/obj/item/stack/cable_coil/Initialize(mapload, new_amount = null, param_color = null)
```
Comment (ike709) [inline code/modules/power/cable.dm]: This incorrect usage of Initialize() is unnecessary.

File: code/game/objects/noose.dm [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+/mob/living/carbon/human/proc/checknoosedrop()
+	if(noosed)
+		for(var/obj/structure/chair/noose/noose in loc)
```
Comment (ike709) [inline code/game/objects/noose.dm]:
```suggestion
		for(var/obj/structure/chair/noose/noose in get_turf(src))
```

File: code/game/objects/noose.dm (hunk 1 above)
Comment (ike709) [reply code/game/objects/noose.dm]: `if(R.result_type == /whatever/noose/path/is)`

File: code/game/objects/noose.dm (hunk 4 above)
Comment (ike709) [reply code/game/objects/noose.dm]: `var/image/over`

Comment (ro5490) [reply code/game/objects/noose.dm]: not sure, was direct ported from hippie, ill fix that though

File: code/game/objects/noose.dm (hunk 5 above)
Comment (ro5490) [reply code/game/objects/noose.dm]: so, change var/noosed to a buckled mobs check instead, got it

File: code/modules/power/cable.dm (hunk 7 above)
Comment (ro5490) [reply code/modules/power/cable.dm]: How do I go about fixing this?

File: code/modules/surgery/bodyparts/head.dm (hunk 6 above)
Comment (ro5490) [reply code/modules/surgery/bodyparts/head.dm]: not exactly sure, was direct ported from hippiecode, ill look into it, any suggestions here?

File: code/modules/power/cable.dm (hunk 7 above)
Comment (ike709) [reply code/modules/power/cable.dm]: Remove it. It's unneeded.

File: code/game/objects/noose.dm (hunk 1 above)
Comment (ro5490) [reply code/game/objects/noose.dm]: ahh cool, Will fo that now

File: code/modules/power/cable.dm (hunk 7 above)
Comment (ro5490) [reply code/modules/power/cable.dm]: so 
```
/obj/item/stack/cable_coil/Initialize(mapload, new_amount = null, param_color = null)
	.=..()
	recipes = list(new/datum/stack_recipe("cable restraints", /obj/item/restraints/handcuffs/cable, 15), new/datum/stack_recipe("noose", /obj/structure/chair/noose, 30, time = 80, one_per_turf = 1, on_floor = 1))

```
to 
```
/obj/item/stack/cable_coil/Initialize()
	.=..()
	recipes = list(new/datum/stack_recipe("cable restraints", /obj/item/restraints/handcuffs/cable, 15), new/datum/stack_recipe("noose", /obj/structure/chair/noose, 30, time = 80, one_per_turf = 1, on_floor = 1))
```
Or just delete the whole block?

File: code/modules/surgery/bodyparts/head.dm (hunk 6 above)
Comment (ro5490) [reply code/modules/surgery/bodyparts/head.dm]: >removed this and the proc itself as its not needed it seems

File: code/game/objects/noose.dm (hunk 5 above)
Comment (ro5490) [reply code/game/objects/noose.dm]: >removed this and the `checknoosedrop()`proc itself as its not needed it seems

Comment (ro5490) [reply code/game/objects/noose.dm]: also updated code to have buckled mob check instead of ifnoosed

File: code/modules/power/cable.dm (hunk 7 above)
Comment (ike709) [reply code/modules/power/cable.dm]: The whole block

**Comment by ro5490** [comment]
Ik the whole visible message section in noose.dm is a mess, tried to fix it and got nowhere cause i got confused for a hot sec, Will continue to fiddle with it but if anyone can fix it up for me or give me pointers I'd appreciate it

Also added a trait check so it doesnt try and oxyloss mobs with no_breath trait

File: code/game/objects/noose.dm [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+		if(buckled_mob.mob_has_gravity())
+			if(!HAS_TRAIT(buckled_mob, TRAIT_NOBREATH))
+				buckled_mob.adjustOxyLoss(5)
+				if(prob(40))
+					buckled_mob.emote("gasp")
+			if(prob(20))
+				var/flavor_text = list("<span class='suicide'>[buckled_mob]'s legs flail for anything to stand on.</span>",\
+										"<span class='suicide'>[buckled_mob]'s hands are desperately clutching the noose.</span>",\
+										"<span class='suicide'>[buckled_mob]'s limbs sway back and forth with diminishing strength.</span>")
+				if(buckled_mob.stat == DEAD)
```
Comment (qwertyquerty) [inline code/game/objects/noose.dm]: Take out the messages for if they are dead. It will just be a spam. Only send a message if they aren't dead.

**Comment by Goshagosha** [comment]
You can behead hanging people and they stay hanging. Looks ugly.

**Pull request merged**